libjcat/jcat-bt-util.c

```diff
@@ -0,0 +1,1260 @@
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "cannot find end of key name");
+		return FALSE;
+	}
+	if (p == signature_start) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "key name cannot be empty");
+		return FALSE;
+	}
+	if (memchr(signature_start, '+', p - signature_start) != NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "key name cannot contain a plus sign");
+		return FALSE;
+	}
+	key_name = g_strndup(signature_start, p - signature_start);
+	if (strcmp(key_name, expected_key_name) != 0) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "unexpected key name; expecting %s but found %s",
+			    expected_key_name,
+			    key_name);
+		return FALSE;
+	}
+	signature_size -= p - signature_start + 1;
+	signature_start = p + 1;
+	for (p = signature_start; p < signature_start + signature_size; ++p) {
+		if (!g_ascii_isalnum(*p) && *p != '+' && *p != '/' && *p != '=') {
+			g_set_error(
+			    error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "incorrect base64 encoding for signature: invalid character 0x%02x",
+			    (unsigned)*p);
+			return FALSE;
+		}
+	}
+	temp_sig = g_strndup(signature_start, signature_size);
+	sig_str = g_base64_decode(temp_sig, &sig_size);
+
+	/* The first four bytes are actually the key hash. */
+	actual_key_hash = sig_str;
+	if (memcmp(expected_key_hash, actual_key_hash, 4) != 0) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "actual key hash does not match expected key hash");
+		return FALSE;
+	}
+
+	if (sig_size - 4 != ED25519_SIGNATURE_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "invalid size for signature; should have size %d but has size %zu",
+			    ED25519_SIGNATURE_SIZE,
+			    sig_size - 4);
+		return FALSE;
+	}
+
+	if (ed25519_sha512_verify(parsed_public_key->data,
+				  g_bytes_get_size(*original_text),
+				  g_bytes_get_data(*original_text, NULL),
+				  sig_str + 4) == 0) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "cannot verify Ed25519 signature");
+		return FALSE;
+	}
+	return TRUE;
+#endif
+}
+
+/*
+ The on-disk structure is:
+
+	<rootDir>/leaves/aa/bb/cc/ddeeff...
+	<rootDir>/leaves/pending/aabbccddeeff...
+	<rootDir>/seq/aa/bb/cc/ddeeff...
+	<rootDir>/tile/<level>/aa/bb/ccddee...
+	<rootDir>/checkpoint
+ */
+
+static gboolean
+create_storage(GFile *storage_dir, GError **error)
+{
+	gboolean ret;
+	g_autoptr(GFile) seq_dir = NULL;
+	g_autoptr(GFile) tile_dir = NULL;
+	g_autoptr(GFile) leaves_dir = NULL;
+	g_autoptr(GFile) pending_dir = NULL;
+
+	ret = g_file_make_directory_with_parents(storage_dir, NULL, error);
+	if (!ret)
+		return FALSE;
+
+	/* Create necessary subdirectories */
+	seq_dir = g_file_get_child(storage_dir, "seq");
+	ret = g_file_make_directory(seq_dir, NULL, error);
+	if (!ret)
+		return FALSE;
```

Comment:
don't use ret, just put the cmd in the if block pls

Reply:
Done.